Generate names for cities

mayant15 opened the pull request:
Solves #22 by generating city names.

**To test:**
- Find a city and go near its centre
- There'll be a blue name tag showing the city's name
- Specify a theme in your culture prefab to customise the generated name

Commit:
Add random names to cities

Commit:
Add random names to cities

Commit:
Add names at world generation

Names are now assigned to sites on world generation as opposed to
on encounter.

Commit:
Add names at world generation

Names are now assigned to sites on world generation as opposed to
on encounter.

Commit:
Clean up and format code

Commit:
Clean up and format code

Cervator commented on src/main/java/org/terasology/dynamicCities/settlements/SettlementEntityManager.java:
```diff
@@ -120,19 +121,21 @@
 
         settlementEntities = settlementCachingSystem.getSettlementCacheEntity();
-        rng = new FastRandom(regionEntityManager.hashCode() & 0x921233);
+        long seed = regionEntityManager.hashCode() & 0x921233;
```
Would it be worth trying to hook into the world seed instead? Also might help to make a variable/constant for the magic string here to explain its meaning (it seems fairly obvious but is a good practice)

syntaxi replied:
I think leaving this one as is is fine.
The names are hooked into the world seed below, this is for the other randoms, like city generation over time. If you hooked it into the seed it wouldn't really matter because all the other factors would be different

Cervator commented on src/main/java/org/terasology/dynamicCities/settlements/SettlementEntityManager.java:
```diff
@@ -120,19 +121,21 @@
         if (!settlementCachingSystem.isInitialised()) {
             return;
-        } else if (settlementCachingSystem.isInitialised() && settlementEntities == null) {
+        } else if (settlementEntities == null) {
```
How come the other check goes away? More curiosity than anything else

mayant15 replied:
Because of the previous line, `if (!settlementCachingSystem.isInitialised()) return`, this check is redundant. I removed it because it came up as a warning.

syntaxi replied:
Ah yup. Looks makes sense to me

Cervator reviewed (approved):
Looks fine to me, although I haven't tested yet pending WIP :-)

Would be nice with some quick instructions of what to look for / how to test when ready.

Commit:
Split up Settlement and Site components

Commit:
Split up Settlement and Site components

Commit:
Allow town theme to be specified in the culture component

mayant15 commented on src/main/java/org/terasology/dynamicCities/population/CultureComponent.java:
```diff
@@ -58,4 +63,17 @@ public float getProcentualOfZone(String zone) {
         return procentuals;
     }
+
+    public TownTheme getTownTheme() {
```
There must be a better way to do this. Ideally it should be a solution which allows devs using DynamicCities to define their own custom themes inside their own modules.

Cervator replied:
Yeah, on top of that `Component` classes aren't actually meant to have code at all in the first place - they're meant to be pure data containers. Long time since I looked but I thought you could register cultures via prefabs?

mayant15 commented on src/main/java/org/terasology/dynamicCities/settlements/SettlementEntityManager.java:
```diff
@@ -200,9 +204,16 @@ private EntityRef createSettlement(EntityRef siteRegion) {
+        SettlementComponent settlementComponent = siteRegion.getComponent(SettlementComponent.class);
+
+        // Generate name
+        TownNameProvider nameProvider = new TownNameProvider(rng.nextLong(), cultureComponent.getTownTheme());
```
Okay so now the name generation is NOT hooked to the world seed. Earlier it happened in the SettlementFacetProvider which had access to the world seed. Now I have to generate it after the CultureComponent is ready.

mayant15 commented:
Updated. Had some questions so I left them as comments :)

Commit:
Add custom themes

Commit:
Code style fixes

Commit:
Restructure SettlementComponent

Commit:
Add default constructor to SettlementComponent

skaldarnar commented on src/main/java/org/terasology/dynamicCities/population/ThemeManager.java:
```diff
@@ -0,0 +1,100 @@
+
+@Share(ThemeManager.class)
+@RegisterSystem(RegisterMode.AUTHORITY)
+public class ThemeManager extends BaseComponentSystem {
```
I'd like to see a short class comment here explaining what this manager is about, and what a _theme_ is supposed to be. Not critical for mergin though...

skaldarnar commented on src/main/java/org/terasology/dynamicCities/population/ThemeManager.java:
```diff
@@ -0,0 +1,100 @@
+    @In
+    AssetManager assetManager;
+
+    private Set<TownNameComponent> themeComponents = new HashSet<>();
```
Why are `themeComponents` actually `TownNameComponents`? Just out of convenience as the town name happens to contain information about the theme as well?

skaldarnar commented on src/main/java/org/terasology/dynamicCities/population/ThemeManager.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Override
+    public void postBegin() {
+        logger.info("Loading default themes...");
```
I wonder whether we should reduce the log levels here... should those `info` logs actually be `debug` instead?

skaldarnar commented on src/main/java/org/terasology/dynamicCities/population/ThemeManager.java:
```diff
@@ -0,0 +1,100 @@
+            }
+        }
+        themeNames += "]";
+        logger.info("Finished loading themes: " + themeComponents.size() + " theme types found: " + themeNames);
```
The code above is just used for generating the log message, isn't it? I think we should remove it. 

The following code using Stream API should yieldthe same result as your iterator-based code. The code is more declarative though which often helps to think more about the _what_ than the _how_ 😉 
```java
themeComponents
  .stream()
  .map(c -> c.name)
  .collect(Collectors.joining(", ", "[", "]")
  .toString()
```

skaldarnar commented on src/main/java/org/terasology/dynamicCities/population/TownNameComponent.java:
```diff
@@ -0,0 +1,28 @@
+    public String name;
+    public List<String> prefixes;
+    public List<String> postfixes;
+    public List<String> nameList;
```
What is `nameList` supposed to be, given there are already `name` and `prefixes` and `postfixes` ? Is it some kind of tag list?

skaldarnar reviewed (approved):
Overall looks good to me. Will test it in game now =)

skaldarnar commented on src/main/java/org/terasology/dynamicCities/population/ThemeManager.java:
```diff
@@ -0,0 +1,100 @@
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterMode;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.namegenerator.town.DebugTownTheme;
```
One thing: is the `namegenerator` a transitive dependency? Do we need to add it to the module dependencies?

skaldarnar commented:
Game crashed right when I discovered the first city.... 
https://pastebin.com/aju33JPY

_Edit: I did not have any culture activated... with MedievalCities in the active modules I can see that the city has a name when looking in the console, but there is no name tag showing in the world_

I would like to avoid a hard crash like the one that happens when you don't active an additional culture module, but the hardening against these case could also happen in another PR...

_Edit 2: Finally got a working version with name tags showing... @Cervator what's the difference between using MedievalCities game mode and just activating the modules on their own? The first works as intended, the second does not show name tags for me..._

![Terasology-190412235253-1681x946](https://user-images.githubusercontent.com/1448874/56068252-70213e00-5d7e-11e9-80dc-e9ac8916f413.png)

Cervator commented:
Bump - @mayant15 you still out there? :-)

Commit:
Refactor code

Commit:
Add NameGenerator to module dependencies

Commit:
Remove sufficient grass condition

The SiteFacetProvider no longer looks for sufficient grass when deciding if
a site is suitable for building a city.

Commit:
Add city centre icon to minimap

Commit:
Fix icon scaling with zoom

Commit:
Add network constraints to CentreOverlay addition

Commit:
Make icon persist when city moves out of view

Commit:
Add javadocs

AndyTechGuy committed:
Add location component to building entities

Commit:
Setup road construction pipeline

 - Created basic container classes
 - Created temporary mock functions
 - Setup SettlementEntityManager to create roads as it creates buildings

Commit:
Calculate and place RoadParcels between settlements

Commit:
Add a rasterizer to place blocks

Commit:
Better place and clear out space for roads

- Buildings of the source will not be built upon roads
- RoadSegments will be laid down better

Commit:
Add gravel blocks to road

Commit:
Add road rasterizer and theme to construction system

Commit:
Create roads from the nearest city only

Commit:
Clean up code

Commit:
Refactor and add javadocs

mayant15 commented:
Closing this one, consider #46 instead.
